### PR TITLE
make target states work together with adaptive voronoi

### PR DESCRIPTION
**Pull request by fabian-paul**
Hi Ali!

Thanks for implementing the the adaptive Voronoi binning!

However I found that the adaptive Voronoi binning doesn't really allow you to properly define target states. 

Here is my suggestion, how this problem could possibly be solved. Comments and critique are welcome.

I suggest to separate the concept of "bins" from the concepts of "states". In the current code base, these two things are a bit entangled. So I just propose to disentangle them a bit. Instead of looking whether a segment ends in a target bit, I check whether it ends in a target state. For this, the target state needs to be able to "tell" whether a given pcoord is inside the state or not. I suggest to add a `__contains__` method to the `TargetState` class with does exactly that. The current implementation of `__contains__`  is just a proof of concept and has to refined, if this whole idea works.

Best,
Fabian

### CODE REVIEW

**Comment by ASinanSaglam** [comment]
Hi Fabian! 

Thank you for coming up with this, I was planning to add this later on as well. I'll give this a run tomorrow and see how it works, this is definitely a much needed option!

**Comment by ASinanSaglam** [comment]
Hi Fabian,

Sorry I didn't respond to this yet, I messed around with the idea a bit but I'm only managing to spend some more time thinking about it just now. 

I really like the idea, in particular since something like this (specifically decoupling target states and bins) will have to be done in the future anyway if we want to support more adaptive binning schemes. I considered hacking it in using a functional/recursive bin mapper (e.g. embedding a voronoi mapper inside a rectilinear map) but I'm not sure if that's a good way of doing it since that will over complicate the updates to the mapper

Basically, I'm just currently not sure what is the best way to achieve this, I'll play around with this idea a bit more. Did you run a test simulation with your implementation?

**Comment by ASinanSaglam** [comment]
I made a change to include a user defined function and played with the ODLD system config to show the idea off. I probably shouldn't have committed the changes to the ODLD example, I'll change those back later. 

Give this a try, from my preliminary test this works for steady state simulations. Essentially what I'm doing is to embed a voronoi mapper within a particular region of the state space and update that particular mapper. The outer mapper allows me to recycle once I hit that boundary and I still keep my adaptive voronoi sampling. 

This needs further testing to ensure it works as intended of course but I think it is at least a temporary solution until we can discuss a more general solution for the targetstate/adaptive binning scheme issues with WESTPA devs.

**Comment by fabian-paul** [comment]
Hi Ali,

Regarding tests: yes I tested my code using my own toy model. I just checked that trajectories that reach the target state get recycled. But I didn't check convergence of the simulation or anything systematic. That would be on my todo list.

About the decision whether the separation of state and bins should be done in a plugin or in the westpa core:
Management of states sounds to me like a core function of westpa. So I would not outsource it to a plugin. 
This all depends of course on where the main development direction of westpa is going. It depends on the anticipated use cases: will 90% of users use 1-D progress coordinates in the future? If so, the effort to change westpa might not pay off. If adaptive Voronoi or WExplore is the typical future use case, it is worthwhile to change westpa. 

As you already wrote, if states are different from bins, we have to find a good way to let the user define the target states. I'm a bit inclined against plugins and user functions, because this over-complicates simple use cases. It's nice that westpa has a central west.cfg configuration file, so we should take advantage of this (it much nicer than searching through various python modules like system.py and the plugin sources to find the correct spot for configuration parameters).

I think that one could add a "target" section to west.cfg that allows to define the target states. Here are some examples, how use case might look like:
```
target:
    condition: "x[0] > 100"
```

```
target:
    reference: "target.npy"
    type: numpy
    condition: "numpy.linalg.norm(x - reference) < 0.3"
```

```
target:
    reference: "target.pdb"
    type: mdtraj
    condition: "mdtraj.rmsd(x, reference) < 0.3"
```

User-provided functions would also be possible for complicated cases.

```
target:
    condition: "user_module.user_function(x)"
```

About your temporary solution: let's hear what the developers says. I just requested access to the mailing list, so I can't see the discussion yet.

Best,
Fabian

**Comment by ASinanSaglam** [comment]
Hi Fabian, 

At the moment the main use case for WESTPA is still 1-2D steady/equilibrium MD simulations. But I think in general allowing it to be used for anything would be the best solution so I definitely am for this type of decoupling. Even if it stays as MD-focused, 1/2D simulations, there are definite benefits to allowing adaptive binning and this question will always come up if we have the target states = bins idea in place. 

> Management of states sounds to me like a core function of westpa. So I would not outsource it to a plugin.

I would agree with this statement which brings me to my current thinking, I think we can get this plugin merged to the main repo regardless of this discussion and then we can make a separate branch to work on the target state issue specifically, depending on how the discussion goes. Which makes more sense to me especially since we'll be modifying core files and probably don't want to couple this plugin branch with core changes (and this plugin should work regardless of those changes I think). Additionally, I think you might want to bring this up as an issue on the [main WESTPA repo](https://github.com/westpa/westpa/issues) if the Google group discussion doesn't seem very active. You want to give it about a couple weeks since one of the main devs is busy for 1-2 weeks. 

While it does complicates things, I don't think there is any harm to include that user defined function to build the bin mapper in the plugin which can be used for hacking steady state in but also you can use it for complicated binning schemes so I think it's nice to have.
